apps/docs/refactor.md

```diff
@@ -0,0 +1,565 @@
+exposure to unencapsulated spaghetti code.
+
+Getting back to "library types", I think there are two different library types
+for which we need to establish well defined patterns.
+
+1. **feature library** - This is a library that provide some meaty feature
+   of a page. It is characterized by the following attributes:
+
+   - maintains UI represented in the DOM (i.e. calls to ReactDOM.render(), or
+     use of jQuery to manipulate the DOM)
+   - interacts with a remote, stateful service (i.e. ajax calls that generate
+     writes to a database)
+   - uses or provides shared application state
+
+   Some examples of "feature libraries" in our code base today would be the code
+   found in `src/applab`, `src/netsim`, and `src/flappy`. Note however that
+   these directories also contain `main.js` files which strictly speaking are
+   not part of the "library" but are actually entry points that get loaded on a
+   particular html page.
+
+2. **capsule library** - This is a library that is wholy encapsulated and
+   decoupled, and which therefore:
+
+   - does not maintain any UI represented in the DOM (react components are OK, but
+     calls to ReactDOM.render() are not)
+   - does not interact with remote, stateful services (ajax calls to read-only
+     services could be OK)
+   - does not use or provide shared application state (at least not in a direct way)
+
+   Some examples of "capsule libraries" in our code base would be
+   `src/MusicController.js`, `src/dom.js`, `src/Sounds.js`.
+
+
+## Capsule Library ##
+
+### Capsule Library Pattern ###
+
+In the case of a relatively simple capsule library that consists of only one
+file, the pattern is pretty simple:
+
+- Export functions and classes that other libraries are allowed to use. Do not
+  change the interfaces of these functions/classes in a way that is not
+  backwards compatible without updating all the callsites.
+- If the library works with internal state, encapsulate that state into a class
+  that must be instantiated by the calling code.
+- If the library works with internal state, but only supports a singleton
+  instance, only export the singleton instance and not the backing class.
+
+For more complicated capsule libraries whose code is spread across multiple
+files the pattern is only a bit more involved:
+
+- Put all the library files into a directory, and expose a single index.js file
+  which exports the functions/classes that other libraries are allowed to
+  use. Calling code imports from the index.js file only. This makes dependency
+  graphs easier to read.
+
+### Capsule Library Rules ###
+
+The rules for capsule libraries are also straightforward:
+
+- Do not depend on feature libraries
+- Do not read from or write to the global window object (accessing standard browser
+  APIs is OK)
+- Do not reference DOM nodes by ID or css selectors that are not passed in to
+  the library explicitly.
```

Comment:
My argument is that if you are constructing dom elements yourself, then you probably have references to those dom elements already. Using a powerful dom traversal library to lookup a dom node that you could have just stored a reference to seems like overkill.

For example, the following seems a bit crazy to me,
```
$(containerEl).append('<div class="msg-element">hello world</div>')
// ... somewhere else in your code ...
$(containerEl).find('.msg-element').html("a new message");
```
when you could just do this instead:
```
var msgElement = $('<div>hello world</div>')[0];
$(containerEl).append(msgElement);
// ... somewhere else in your code ...
$(msgElement).html("a new message");
```

The beneficial side effects of the latter example are that you aren't dirtying up the global css class namespace, and you are making it harder for other code to (un)intentionally monkey around with the dom you created.

In short, using selectors to access "local" dom nodes is the same as putting all your local variables on `window` and referencing them all through `window` when you could just as easily have them enclosed in a scope.